docs/source/asr/datasets.rst

```diff
@@ -633,6 +633,135 @@ Some other Lhotse related arguments we support:
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Combining a large number of datasets and defining weights for them can be tricky.
+We offer an extended configuration format that allows you to explicitly define datasets,
+dataset groups, and their weights either inline in the experiment configuration,
+or as a path to a separate YAML file.
+
+In addition to the features above, this format introduces a special ``tags`` dict-like field.
+The keys and values in ``tags`` are automatically attached to every sampled example, which
+is very useful when combining multiple datasets with different properties.
+The dataset class which converts these examples to tensors can partition the mini-batch and apply
+different processing to each group.
+For example, you may want to construct different prompts for the model using metadata in ``tags``.
+
+Example 1. Combine two datasets with equal weights and attach custom metadata in ``tags`` to each cut:
+
+.. code-block:: yaml
+
+    input_cfg:
+      - type: nemo_tarred
+        manifest_filepath: /path/to/manifest__OP_0..512_CL_.json
+        tarred_audio_filepath: /path/to/tarred_audio/audio__OP_0..512_CL_.tar
+        weight: 0.4
+        tags:
+          lang: en
+          pnc: no
+      - type: nemo_tarred
+        manifest_filepath: /path/to/other/manifest__OP_0..512_CL_.json
+        tarred_audio_filepath: /path/to/other/tarred_audio/audio__OP_0..512_CL_.tar
+        weight: 0.6
+        tags:
+          lang: pl
+          pnc: yes
+
+Example 2. Combine multiple (4) datasets, corresponding to different tasks (ASR, AST).
+Each task gets its own group and its own weight.
+Then within each task, each dataset get its own within-group weight as well.
+The final weight is the product of outer and inner weight:
+
+.. code-block:: yaml
+
+    input_cfg:
+      - type: group
+        weight: 0.7
+        tags:
+          task: asr
+        input_cfg:
+          - type: nemo_tarred
+            manifest_filepath: /path/to/asr1/manifest__OP_0..512_CL_.json
+            tarred_audio_filepath: /path/to/tarred_audio/asr1/audio__OP_0..512_CL_.tar
+            weight: 0.6
+            tags:
+              source_lang: en
+              target_lang: en
+          - type: nemo_tarred
+            manifest_filepath: /path/to/asr2/manifest__OP_0..512_CL_.json
+            tarred_audio_filepath: /path/to/asr2/tarred_audio/audio__OP_0..512_CL_.tar
+            weight: 0.4
+            tags:
+              source_lang: pl
+              target_lang: pl
+      - type: group
+        weight: 0.3
+        tags:
+          task: ast
+        input_cfg:
+          - type: nemo_tarred
+            manifest_filepath: /path/to/ast1/manifest__OP_0..512_CL_.json
+            tarred_audio_filepath: /path/to/ast1/tarred_audio/audio__OP_0..512_CL_.tar
+            weight: 0.2
+            tags:
+              source_lang: en
+              target_lang: pl
+          - type: nemo_tarred
```

Comment:
Another concern is - this config is fine for first run training, but once a model is trained it's config is set in stone. During fine-tuning, the user will need to heavily modify the config (and be knowledgeable of what to change) in order to finetune the model. We need clear explanation and guidelines of what they need to change.

Reply:
> During fine-tuning, the user will need to heavily modify the config (and be knowledgeable of what to change) in order to finetune the model. We need clear explanation and guidelines of what they need to change.

Yes, I agree. In most cases it will be straightforward (set `input_cfg=None` and `manifest_filepath=my_manifest.json` or similar) but we need to communicate it well. I'll add this to the docs and try to make it visible.

>  once a model is trained it's config is set in stone

I think about it differently, from a user POV the data config is irrelevant as they won't have the same data/paths as the model creator. From a model creator POV this config is going to keep changing: new datasets, different data variants, different data weights. That's why we support keeping different configuration variants in separate YAML files and just providing a path here (it also auto-hides the data config from the users as they will only see `input_cfg: my_data_conf.yaml` in the released `.nemo` file).

Reply:
This statement isn't really true. While yes data loader can be different after training, till now the user has to make at most 2 changes to get training on single GPU - change manifest file path and set is_tarred: false. Rest all works mostly. 

That's no longer the card with this config and it's complicated nest

Reply:
It's still two flags: `input_cfg=None` and `manifest_filepath=<path>` (is_tarred is not needed with `use_lhotse=True`). If you mean the batching settings (batch size, duration, bucketing, etc) they can stay as they are (they are not a part of `input_cfg` if that's what you were thinking).